fog/sample-paykit/src/cached_tx_data/memo_handler.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+//! A memo handler object which processes memos, for use in integration tests
+
+use displaydoc::Display;
+use mc_account_keys::{AccountKey, AddressHash, PublicAddress, CHANGE_SUBADDRESS_INDEX};
+use mc_common::logger::{log, Logger};
+use mc_crypto_keys::{KeyError, RistrettoPublic};
+use mc_transaction_core::{get_tx_out_shared_secret, subaddress_matches_tx_out, tx::TxOut};
+use mc_transaction_std::{MemoDecodingError, MemoType};
+use std::{collections::HashMap, convert::TryFrom};
+
+/// A handler object that holds a contacts list and tries to recieve and
+/// authenticate memos. It provides the "get_last_memo" function which can be
+/// used to check what the last memo recieved was, and if there were validation
+/// errors.
+///
+/// This is useful for test code.
+#[derive(Debug, Clone)]
+pub struct MemoHandler {
+    contacts: HashMap<AddressHash, PublicAddress>,
+    last_memo: Result<Option<MemoType>, MemoHandlerError>,
+    logger: Logger,
+}
+
+impl MemoHandler {
+    /// Make a new memo handler with a given set of contacts
+    pub fn new(address_book: Vec<PublicAddress>, logger: Logger) -> Self {
```

Comment:
Nit: `&[PublicAddress]`?

Reply:
i guess i prefer to take it by value, since it avoids a copy? why do you have this nit?

Reply:
Disregard, I missed the part where it grabs `addr`.